lib/pure/asyncdispatch.nim

```diff
@@ -1559,8 +1559,8 @@ proc drain*(timeout = 500) =
   ## Waits for completion events and processes them. Raises ``ValueError``
   ## if there are no pending operations. In contrast to ``poll`` this
   ## processes as many events as are available.
-  if runOnce(timeout):
-    while hasPendingOperations() and runOnce(0): discard
+  if runOnce(timeout) or hasPendingOperations():
+    while hasPendingOperations() and runOnce(timeout): discard
```

Comment:
I changed this from `runOnce(0)` to `runOnce(timeout)` because the `0` is a sentinel value for "sleep forever, until some selector is woken up". But a non-zero timeout will have very different behavior.

For the common case, I don't think this will matter too much, but If somebody passed a particular timeout into this function, they probably expected a certain behavior. Forcing `runOnce(0)` may produce unexpected results in that case.

Reply:
> because the 0 is a sentinel value for "sleep forever, until some selector is woken up".

That's not correct, you're mistaking `0` for `-1`. For `0`, `runOnce` will return immediately. For `-1`, `runOnce` will wait forever until some FDs get events.

Reply:
>That's not correct, you're mistaking 0 for -1. For 0, runOnce will return immediately. For -1, runOnce will wait forever until some FDs get events.

ah, yes! That makes sense.

Though, I still feel like the right thing to do here is use the user specified timeout value.

Reply:
I don't think so, because this will mean that ``drain`` might block for longer than ``timeout``.

Reply:
> 
> 
> I don't think so, because this will mean that `drain` might block for longer than `timeout`.

I don't understand what you mean by this? 

The original code (as it currently exists on develop) looks like this:
```
  if runOnce(timeout):
    while hasPendingOperations() and runOnce(0): discard
```

We call `runOnce` without a timeout for the first iteration, waiting for the timeout, and then call `runOnce` with `0`, causing it to spin loop as fast as possible.

Essentially it's, "block for the timeout the first time, but then spin as fast as possible after."
That is really weird behavior imo. Not what I expect to happen anyway.

Reply:
I think the assumption is that this while loop won't take many loops to finish. From that perspective it actually shouldn't matter what timeout you give it. 

Maybe @Araq can give his perspective, since he added this procedure...

Reply:
> Essentially it's, "block for the timeout the first time, but then spin as fast as possible after."
That is really weird behavior imo. Not what I expect to happen anyway.

It's exactly how ``drain`` should look like IMO. It's very rarely that you need it, but what's questionable about the design? And why does it need to be changed? Sorry, I didn't read everything in this thread...

Reply:
@Araq:
Quick recap. I made two changes to `drain`:

1.) I changed the if condition from `if runOnce(timeout)` to `if runOnce(timeout) or hasPendingOperations():`

It's possible for outstanding pending operations to still exist even if `runOnce` returns false, and must be accounted for. This was causing bugs on the slow windows VMs used in the CI.
This change is not so controversial, IMO.

2.) In the while loop I changed the `runOnce(0)` to `runOnce(timeout)`

I have two reasons for this change:
- Consistency: Every other proc in the asyncdispatch api treats the timeout as the value for waiting on the selector `poll`, not an absolute timeout. This is the one proc that is different.

- Going back to point 1 above, there may be pending selector events that are very slow to be ready. For example, imagine a slow network causing a socket to be waiting on the order of seconds, or a slow VM creating a delayed hard drive write.

In the current implementation, the loop might spin thousands of times, burning the cpu and not actually getting any real work done. It's not safe to assume that the loop will complete in only a few iterations.

It's much more efficient to wait on the timeout every iteration.

IMO, it's more consistent with what I expect to happen as well.

Reply:
Both changes are fine but need to be mentioned in the changelog.